controller/registry/dockerclient.py

```diff
@@ -0,0 +1,116 @@
+        self.build(source, config, name, tag)
+
+        # push the image to deis-registry
+        self.push("{}/{}".format(self.registry, name), tag)
+
+    def build(self, source, config, repo, tag):
+        """Add a "last-mile" layer of environment config to a Docker image for deis-registry."""
+        self._check_blacklist(repo)
+        env = ' '.join("{}='{}'".format(
+            k, v.encode('unicode-escape').replace("'", "\\'")) for k, v in config.viewitems())
+        dockerfile = "FROM {}\nENV {}".format(source, env)
+        f = io.BytesIO(dockerfile.encode('utf-8'))
+        target_repo = "{}/{}:{}".format(self.registry, repo, tag)
+        logger.info("Building Docker image {}".format(target_repo))
+        with SimpleFlock(self.FLOCKFILE, timeout=1200):
+            stream = self.client.build(fileobj=f, tag=target_repo, stream=True, rm=True)
+            log_output(stream)
+
+    def pull(self, repo, tag):
+        """Pull a Docker image into the local storage graph."""
+        self._check_blacklist(repo)
+        logger.info("Pulling Docker image {}:{}".format(repo, tag))
+        with SimpleFlock(self.FLOCKFILE, timeout=1200):
+            stream = self.client.pull(repo, tag=tag, stream=True, insecure_registry=True)
+            log_output(stream)
+
+    def push(self, repo, tag):
+        """Push a local Docker image to a registry."""
+        logger.info("Pushing Docker image {}:{}".format(repo, tag))
+        stream = self.client.push(repo, tag=tag, stream=True, insecure_registry=True)
+        log_output(stream)
+
+    def tag(self, image, repo, tag):
+        """Tag a local Docker image with a new name and tag."""
+        self._check_blacklist(repo)
+        logger.info("Tagging Docker image {} as {}:{}".format(image, repo, tag))
+        if not self.client.tag(image, repo, tag=tag, force=True):
+            raise docker.errors.DockerException("tagging failed")
+
+    def _check_blacklist(self, repo):
+        # Blacklist any deis/* repositories. This is not foolproof, but should prevent
+        # likely "deis pull" mistakes.
+        # NOTE: please keep this list up to date!
+        blacklisted = [
+            'builder', 'cache', 'controller', 'database', 'logger', 'logspout',
+            'publisher', 'registry', 'router', 'store-admin', 'store-daemon',
+            'store-gateway', 'store-metadata', 'store-monitor', 'swarm', 'mesos-master',
+            'mesos-marathon', 'mesos-slave', 'zookeeper',
+        ]
+        if any("deis/{}".format(c) in repo for c in blacklisted):
+            raise PermissionDenied(
+                "Can't pull a repository that could collide with a Deis component")
+
+
+def log_output(stream):
+    for chunk in stream:
+        logger.debug(chunk)
+        # error handling requires looking at the response body
+        if '"error"' in chunk.lower():
+            raise docker.errors.DockerException(chunk)
+
+
+def publish_release(source, config, target, deis_registry):
```

Comment:
I was trying to minimize code change elsewhere by preserving a function as the main interface, but your suggestion would also work and might conserve resources. I'll give it a try.